cmd/adapter/adapter.go

```
@@ -0,0 +1,430 @@
+	if err != nil {
+		return err
+	}
+	return err
+}
+
+func (ad *Adapter) Close() error {
+	var err error
+	if !utils.WaitWaitGroup(&ad.waitWg, time.Minute) {
+		err = errors.New("close timeout")
+	}
+	ad.logger.Info("Closed, err=", err)
+	return nil
+}
+
+//===================== adapter.api =====================
+
+func (ad *Adapter) runServeAPI(ctx context.Context) {
+	ad.logger.Info("Running serve API on ", ad.cfg.Gravity.ApiListenAddr)
+	ad.waitWg.Add(1)
+	go func() {
+		srv := ad.serveAPI(ctx)
+		select {
+		case <-ctx.Done():
+			sctx, _ := context.WithTimeout(context.Background(),
+				time.Second*time.Duration(5))
+			_ = srv.Shutdown(sctx)
+		}
+		ad.logger.Warn("Serving API stopped.")
+		ad.waitWg.Done()
+	}()
+}
+
+func (ad *Adapter) serveAPI(ctx context.Context) *http.Server {
+	router := httprouter.New()
+	router.GET("/v1/log", ad.makeHandlerWithCtx(ctx, ad.logHandler))
+	router.GET("/v1/download", ad.makeHandlerWithCtx(ctx, ad.downloadHandler))
+
+	srv := &http.Server{Addr: ad.cfg.Gravity.ApiListenAddr, Handler: router}
+	go func() {
+		if err := srv.ListenAndServe(); err != http.ErrServerClosed {
+			ad.logger.Fatal("serveAPI(): %s", err)
+		}
+	}()
+	return srv
+}
+
+func (ad *Adapter) logHandler(ctx context.Context, rw http.ResponseWriter, rq *http.Request, p httprouter.Params) error {
+	ad.logger.Info("log(): Request=", rq)
+	qQuery := strings.TrimSpace(rq.URL.Query().Get("query"))
+	qLimit := strings.TrimSpace(rq.URL.Query().Get("limit"))
+
+	var (
+		err error
+	)
+
+	limit := lgLimitLines
+	if qLimit != "" {
+		limit, err = strconv.Atoi(qLimit)
+		if err != nil || limit < 0 {
+			limit, err = lgLimitLines, nil
+		}
+	}
+
+	qr := &api.QueryRequest{
+		Query: buildLql(qQuery, ad.cfg.Logrange.Partition, limit, -limit),
+		Pos:   "tail", Offset: -limit, Limit: limit,
+	}
+
+	ad.logger.Info("log(): Query=", qr.Query)
+	rs := &api.QueryResult{}
+	err = ad.lCli.Query(ctx, qr, rs)
+	if err == nil {
+		grLogEntries, err := ad.toGrLogEntries(rs.Events)
```

This error is ignored. (And it won't be logged below because it overshadows err from parent context). Same for below.

---

applied (not yet merged to this branch)